xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -634,59 +651,97 @@ boolean hasResult() {
+          }
+        }
+      } finally {
+        if (!pendingProcess) {
+          maybeCompletedHandleResourceUpdate(serverInfo);
         }
       }
     }
 
     void onAbsent() {
-      if (respTimer != null && respTimer.isPending()) {  // too early to conclude absence
-        return;
-      }
+      boolean pendingProcess = false;
+      try {
+        if (respTimer != null && respTimer.isPending()) {  // too early to conclude absence
+          return;
+        }
 
-      // Ignore deletion of State of the World resources when this feature is on,
-      // and the resource is reusable.
-      boolean ignoreResourceDeletionEnabled =
-          serverInfo != null && serverInfo.ignoreResourceDeletion();
-      if (ignoreResourceDeletionEnabled && type.isFullStateOfTheWorld() && data != null) {
-        if (!resourceDeletionIgnored) {
-          logger.log(XdsLogLevel.FORCE_WARNING,
-              "xds server {0}: ignoring deletion for resource type {1} name {2}}",
-              serverInfo.target(), type, resource);
-          resourceDeletionIgnored = true;
+        // Ignore deletion of State of the World resources when this feature is on,
+        // and the resource is reusable.
+        boolean ignoreResourceDeletionEnabled =
+            serverInfo != null && serverInfo.ignoreResourceDeletion();
+        if (ignoreResourceDeletionEnabled && type.isFullStateOfTheWorld() && data != null) {
+          if (!resourceDeletionIgnored) {
+            logger.log(XdsLogLevel.FORCE_WARNING,
+                "xds server {0}: ignoring deletion for resource type {1} name {2}}",
+                serverInfo.target(), type, resource);
+            resourceDeletionIgnored = true;
+          }
+          return;
         }
-        return;
-      }
 
-      logger.log(XdsLogLevel.INFO, "Conclude {0} resource {1} not exist", type, resource);
-      if (!absent) {
-        data = null;
-        absent = true;
-        metadata = ResourceMetadata.newResourceMetadataDoesNotExist();
-        for (ResourceWatcher<T> watcher : watchers) {
-          watcher.onResourceDoesNotExist(resource);
+        logger.log(XdsLogLevel.INFO, "Conclude {0} resource {1} not exist", type, resource);
+        if (!absent) {
+          data = null;
+          absent = true;
+          metadata = ResourceMetadata.newResourceMetadataDoesNotExist();
+          AtomicInteger fanOutCount = new AtomicInteger(watchers.size());
+          if (fanOutCount.get() > 0) {
+            pendingProcess = true;
+          }
+          for (ResourceWatcher<T> watcher : watchers.keySet()) {
+            watchers.get(watcher).execute(() -> {
+              try {
+                watcher.onResourceDoesNotExist(resource);
+              } finally {
+                if (fanOutCount.decrementAndGet() == 0) {
+                  maybeCompletedHandleResourceUpdate(serverInfo);
+                }
+              }
+            });
+          }
+        }
+      } finally {
+        if (!pendingProcess) {
```

Comment:
This approach does work, but you might find this approach easier:

```java
private void handleResourceUpdate(...) {
  ...
  serverChannelMap.get(args.serverInfo).flowControlWindow().set(1);
  ...
  maybeCompletedHandleResourceUpdate(args.serverInfo);
}

// elsewhere, just before moving processing to another thread:
          for (ResourceWatcher<T> watcher : watchers.keySet()) {
            serverChannelMap.get(serverInfo).flowControlWindow().incrementAndGet(); // THIS
            watchers.get(watcher).execute(() -> {
              try {
                watcher.onResourceDoesNotExist(resource);
              } finally {
                maybeCompletedHandleResourceUpdate(serverInfo); // And then unconditional
              }
            });
         }
```

It can then be much easier to verify those are paired up, as they are close together. There is some risk, as if you forget to manage the integer, flow control will just ignore that callback completely (vs a hang).